crontabber/tests/test_crontabber.py

```diff
@@ -2149,6 +2150,105 @@ def test_job_list_separated_sloppily(self):
             ok_('Ran FooJob' in infos, infos)
             ok_('Ran BarJob' in infos, infos)
 
+    def test_run_two_jobs_almost_simultaneously_first_time(self):
```

Comment:
Yes, there is a concurrency problem. 

The right thing to do is to have every thread/process that's interested in *reading* and *using* the task information to get exclusive access. It is possible to change our transaction mode for crontabber to be SERIALIZABLE, but you'd still need to figure out how to handle the locking failures.

(The alternative (longer term) solution is to use a queue where the concurrency problem gets offloaded to the queue. I realize that's problematic because of the timing issues, but it is a feature that I think we might need in RelEng shortly.)

In any case, this blog post might be helpful in explaining the how of what needs to happen: http://binarybuffer.com/2012/12/grails-locking-rows-for-update-in-postgresql-with-nowait

It describes pretty much the same problem that you're addressing now.

This is another description of managing a queue in Postgres: http://www.depesz.com/2013/08/30/pick-a-task-to-work-on/

It's a good description of the problem as well, providing some alternative solutions. Depesz throws out the option of using NO WAIT because he doesn't want to manage the application-level logic (haha). I think *we* do want to manage this in the application.